File: app/src/main/java/quintype/com/templatecollectionwithrx/models/story/StorySearchResult.java
package quintype.com.templatecollectionwithrx.models.story;

import com.google.gson.annotations.SerializedName;

import java.util.Collections;
import java.util.List;

/**
 * An object instance representing SearchResult
 *
 * @author Madhu madhu@quintype.com
 */
public class StorySearchResult {

    /**
     * An object instance representing SearchResults
     *
     * @author Madhu madhu@quintype.com
     */
    public static class Results {
        @SerializedName("results")
        private StorySearchResult result;

        public StorySearchResult result() {
            return result;
        }

        @Override
        public String toString() {
            return "SearchResults{" +
                    "result=" + result +
                    '}';
        }
    }


    @SerializedName("term")
    private String term;
    @SerializedName("from")
    private int from;
    @SerializedName("size")
    private int size;
    @SerializedName("total")
    private int total;
    @SerializedName("stories")
    private List<Story> stories = Collections.emptyList();

    public String term() {
        return term;
    }

    public int from() {
        return from;
    }

    public int size() {
        return size;
    }

    public int total() {
        return total;
    }

    /**
     * @return List of stories found by search {@link #term()}
     */
    public List<Story> stories() {
        return Collections.unmodifiableList(stories);
    }

    @Override
    public String toString() {
        return "SearchResult{" +
                "term=" + term +
                ", from=" + from +
                ", size=" + size +
                ", total=" + total +
                ", stories=" + stories +
                '}';
    }
}

File: app/src/main/java/quintype/com/templatecollectionwithrx/models/story/StorySearchResult.kt
package quintype.com.templatecollectionwithrx.models.story

import android.os.Parcel
import android.os.Parcelable
import com.google.gson.annotations.SerializedName
import java.util.*

/**
 * Created TemplateCollectionWithRx by rakshith on 9/4/18.
 */

class StorySearchResult() : Parcelable {
    override fun writeToParcel(dest: Parcel?, flags: Int) {
        dest?.writeString(this.term)
        dest?.writeInt(this.from)
        dest?.writeInt(this.size)
        dest?.writeInt(this.total)
        dest?.writeTypedList(stories)
    }

    override fun describeContents(): Int {
        return 0
    }

    @SerializedName("term")
    private var term: String? = null
    @SerializedName("from")
    private var from: Int = 0
    @SerializedName("size")
    private var size: Int = 0
    @SerializedName("total")
    private var total: Int = 0
    @SerializedName("stories")
    private var stories = emptyList<Story>()

    constructor(parcel: Parcel) : this() {
        this.term = parcel.readString()
        this.from = parcel.readInt()
        this.size = parcel.readInt()
        this.total = parcel.readInt()
        this.stories = parcel.createTypedArrayList(Story.CREATOR)
    }

    /**
     * An object instance representing SearchResults
     *
     * @author Madhu madhu@quintype.com
     */
    class Results {
        @SerializedName("results")
        private val result: StorySearchResult? = null

        fun result(): StorySearchResult? {
            return result
        }

        override fun toString(): String {
            return "SearchResults{" +
                    "result=" + result +
                    '}'.toString()
        }
    }


    /**
     * @return List of stories found by search [.term]
     */
    fun stories(): List<Story> {
        return Collections.unmodifiableList(stories)
    }

    override fun toString(): String {
        return "SearchResult{" +
                "term=" + term +
                ", from=" + from +
                ", size=" + size +
                ", total=" + total +
                ", stories=" + stories +
                '}'.toString()
    }

    companion object CREATOR : Parcelable.Creator<StorySearchResult> {
        override fun createFromParcel(parcel: Parcel): StorySearchResult {
            return StorySearchResult(parcel)
        }

        override fun newArray(size: Int): Array<StorySearchResult?> {
            return arrayOfNulls(size)
        }
    }
}
